kube/configmap.go

```diff
@@ -1,135 +1,125 @@
-	for {
-		err := cmw.watchConfigMaps(resultCH)
-		if err == ErrClosedChannel {
-			log.Info("closed channel event, restart watch")
-			continue
-		} else {
-			return err
-		}
+func (c *ConfigMapWatcher) update(newObj, oldObj interface{}) {
+	newCM, ok := newObj.(*v1.ConfigMap)
+	if !ok {
+		log.Errorf("Failed to get new ConfigMap object")
+		return
 	}
+
+	c.configs <- configMapToEgressConfig(newCM)
 }
-func (cmw *ConfigMapWatcher) watchConfigMaps(resultCH chan<- map[string][]string) error {
 
-	w := cmw.client.CoreV1().ConfigMaps(cmw.ns)
-	opts := cmw.opts
-	opts.Watch = true
-	watcher, err := w.Watch(opts)
-	if err != nil {
-		return err
+func (c *ConfigMapWatcher) del(obj interface{}) {
+	cm, ok := obj.(*v1.ConfigMap)
+	if !ok {
+		log.Errorf("Failed to get ConfigMap object")
+		return
 	}
 
-	evCH := watcher.ResultChan()
-	defer log.Infoln("Watcher: stopped to watch ConfigMaps")
-
-	for {
-		log.Debug("begin for")
-		select {
-		case ev, okCH := <-evCH:
-			if !okCH {
-				return ErrClosedChannel
-			}
-			cmap, ok := ev.Object.(*v1.ConfigMap)
-			if !ok {
-				log.Errorf("Failed to cast event to ConfigMap %v %v, raw: %+v", ev.Type, ev.Object, ev)
-				res, err := cmw.client.CoreV1().ConfigMaps(cmw.ns).List(meta_v1.ListOptions{})
-				log.Infof("res: %v, err: %v", res, err)
-				return err
-			}
-			results := make(map[string][]string)
-			k := fmt.Sprintf("%s/%s", cmap.Namespace, cmap.Name)
-
-			switch ev.Type {
-			case watch.Deleted:
-				log.Infof("Watcher: ConfigMap event %s: %s", ev.Type, cmap.Name)
-				results[k] = []string{}
-			case watch.Error:
-				log.Errorf("Watcher: ConfigMap event %s: %s - ignore", ev.Type, cmap.Name)
-				time.Sleep(10 * time.Second)
-				continue
-			case watch.Added:
-				fallthrough
-			case watch.Modified:
-				results[k] = []string{}
-				log.Infof("Watcher: ConfigMap event %s: %s", ev.Type, cmap.Name)
-				for _, v := range cmap.Data {
-					results[k] = append(results[k], v)
-				}
-			}
-			log.Debugf("Watcher: got tuples: %v", results)
-			resultCH <- results
-		case <-cmw.quitCH:
-			watcher.Stop()
-			return nil
-		}
+	c.configs <- provider.EgressConfig{
+		Resource: provider.Resource{
+			Name:      cm.Name,
+			Namespace: cm.Namespace,
+		},
 	}
 }
 
-func (cmw *ConfigMapWatcher) ListConfigMaps() (map[string][]string, error) {
-	res := make(map[string][]string)
-	cmaps, err := cmw.listConfigMaps()
-	if err != nil {
-		return nil, err
-	}
-
-	for _, cmap := range cmaps.Items {
-		k := fmt.Sprintf("%s/%s", cmap.Namespace, cmap.Name)
-		for _, v := range cmap.Data {
-			res[k] = append(res[k], v)
+func configMapToEgressConfig(cm *v1.ConfigMap) provider.EgressConfig {
+	ipAddresses := make(map[string]struct{})
+	for key, cidr := range cm.Data {
+		_, ipnet, err := net.ParseCIDR(cidr)
+		if err != nil {
+			log.Errorf("Failed to parse CIDR %v from %s in ConfigMap %s/%s", cidr, key, cm.Namespace, cm.Name)
+			continue
 		}
+		ipAddresses[ipnet.String()] = struct{}{}
```

Comment:
a TODO that you do not have to solve would be to merge all CIDR which are contained by others, for example:
- 1.2.3.0/24 is in the set
- 1.2.0.0/16 is added, which should replace 1.2.3.0/24 to reduce the number of routes